design-proposals/vm-memory-snapshot.md

```diff
@@ -0,0 +1,151 @@
+
+## API
+
+### New volume source
+
+This volume source will wrap a PVC and will be called memoryDumpVolumeSource. It should be clear that this volume source will *not* have a matching disk and should be handled appropriately. It will be mounted to a hard coded mounting point in the virt launcher pod.
+The definition look like this:
+
+```golang
+// MemoryDumpVolumeSource represents a volume without a disk that will store a future memory dump
+// this volume is only mounted to the virt launcher pod
+type MemoryDumpVolumeSource struct {
+    PersistentVolumeClaim *v1.PersistentVolumeClaimVolumeSource `json:"persistentVolumeClaim,omitempty"`
+}
+```
+
+### Add the new volume source to a VM
+
+#### Add to VM yaml
+
+One way to add this volume source should be by adding it to the vm yaml for example:
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: my-vm
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        devices:
+          disks:
+          - disk:
+              bus: virtio
+            name: containerdisk
+          - disk:
+              bus: virtio
+            name: cloudinitdisk
+        resources:
+          requests:
+            memory: 512Mi
+      terminationGracePeriodSeconds: 0
+      volumes:
+      - containerDisk:
+          image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+        name: containerdisk
+      - memoryDump:
+          persistentVolumeClaim:
+              claimName: memory-dump
+        name: disk1
+      - cloudInitNoCloud:
+          userData: |-
+            #cloud-config
+            password: fedora
+            chpasswd: { expire: False }
+        name: cloudinitdisk
+```
+
+This can be added before creating the VM, or if the VM already exists need to add and restart the VM to make the pvc mounted.
+
+#### Add memory dump volume as a hotplug
+
+Use the hot-plug API to add a memory dump volume by adding a flag to the `virtctl addVolume` command:
+
+`$ virtctl addvolume my-vm --volume-name=memory-dump --memoryDumpVolume`
+
+This will hotplug the pvc `memory-hotplug` as a memoryDumpVolumeSource to the virt launcher pod, without mounting it to the VMI.
+De-facto the command will call the addvolume subresource as done in hot-plug, will add the volume to the VM volume spec and to the VMI volumes Spec which should add the volume to the VMI volume status as a memoryDump volume, which should trigger the mount to the virt-launcher pod.
+
+### Get memory dump
+#### trigger memory dump
+After adding the pvc to which the memory dump should be saved to, the user can trigger the memory dump in two ways:
+
+1. Run a virtctl command to get a memory dump. This new API will look as follows:
+
+`$ virtctl memoryDump my-vm`
+
+2. Get a memory dump during Online VMSnapshot. This will be done by adding to the VMSnapshot yaml a parameter to indicate that we also want a memory dump. For example:
```

Comment:
Do we need to account for devices?

Reply:
not sure what you mean

Reply:
If a mediated device is attached to the VM and we take an Online Snapshot, do we get the device's memory?